jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/actions/DeployServerlessApplicationAction.kt

```diff
@@ -15,26 +18,50 @@ class DeployServerlessApplicationAction : AnAction(
         message("serverless.application.deploy"),
         null,
         AwsIcons.Resources.LAMBDA_FUNCTION) {
-    private var templateYamlRegex = Regex("template\\.y[a]?ml", RegexOption.IGNORE_CASE)
+    private val templateYamlRegex = Regex("template\\.y[a]?ml", RegexOption.IGNORE_CASE)
 
-    override fun actionPerformed(e: AnActionEvent?) {
+    override fun actionPerformed(e: AnActionEvent) {
 
-        val project = e?.getRequiredData(PlatformDataKeys.PROJECT) ?: throw Exception("Unable to determine project")
+        val project = e.getRequiredData(PlatformDataKeys.PROJECT)
 
-        val virtualFiles = e.getData(PlatformDataKeys.VIRTUAL_FILE_ARRAY) ?: throw Exception("Could not detect template file")
-        val samTemplateFile = virtualFiles[0]
+        val samTemplateFile = getSamTemplateFile(e) ?: throw Exception("Could not detect template file")
         val template = CloudFormationTemplate.parse(project, samTemplateFile)
 
         // TODO : Validate the template file (this is likely too slow to do in update())
 
         DeployServerlessApplicationDialog(project, template.parameters()).show()
     }
 
-    override fun update(e: AnActionEvent?) {
+    override fun update(e: AnActionEvent) {
         super.update(e)
 
-        val virtualFiles = e?.getData(PlatformDataKeys.VIRTUAL_FILE_ARRAY)
+        e.presentation.isVisible = getSamTemplateFile(e) != null
```

Comment:
@abrooksv can we add our own `DataKey` and will it get persisted across between `update` and `performAction`?

We could save ourselves a bit of hassle by doing the file detection once in the `update`, and then storing it as a data-key that the `performAction` then uses.

Minor optimization.

Reply:
Nah, since this action isn't a DataProvider nor is it bound to a UI element as a DataProvider

This pattern looks to be pretty common actually. `update` has to be fast, `performAction` can be a bit slower